File: ios/localasset-core/Sources/LocalAssetCore/registry/InMemoryResourceRegistry.swift
import Foundation

public class InMemoryResourceRegistry: ResourceRegistry {
    private var items: [String: ResourceDescriptor] = [:]
    private let lock = NSLock()
    private let nowMillis: () -> Int64

    public init(nowMillis: @escaping () -> Int64 = currentMillis) {
        self.nowMillis = nowMillis
    }

    public func register(descriptor: ResourceDescriptor) throws {
        try lock.withLock {
            guard items[descriptor.id] == nil else {
                throw ResourceException(.resolutionError, "duplicate resource id: \(descriptor.id)")
            }
            items[descriptor.id] = descriptor
        }
    }

    public func lookup(id: String) -> RegistryLookupResult {
        lock.withLock {
            guard let d = items[id] else { return .missing }
            guard let ttl = d.ttlMillis else { return .hit(d) }
            return d.createdAtMillis + ttl <= nowMillis() ? .expired : .hit(d)
        }
    }

    public func remove(id: String) {
        lock.withLock { _ = items.removeValue(forKey: id) }
    }

    public func cleanup(nowMillis: Int64) {
        lock.withLock {
            items = items.filter { _, d in
                d.ttlMillis.map { d.createdAtMillis + $0 > nowMillis } ?? true
            }
        }
    }
}

File: android/localasset-core/src/main/kotlin/io/github/xesam/android/localasset/core/registry/InMemoryResourceRegistry.kt
package io.github.xesam.android.localasset.core.registry

import io.github.xesam.android.localasset.core.api.RegistryLookupResult
import io.github.xesam.android.localasset.core.api.ResourceRegistry
import io.github.xesam.android.localasset.core.error.ResourceErrorCategory
import io.github.xesam.android.localasset.core.error.ResourceException
import io.github.xesam.android.localasset.core.model.ResourceDescriptor
import java.util.concurrent.ConcurrentHashMap

class InMemoryResourceRegistry(
    private val nowMillis: () -> Long = System::currentTimeMillis,
) : ResourceRegistry {
    private val items = ConcurrentHashMap<String, ResourceDescriptor>()

    override fun register(descriptor: ResourceDescriptor) {
        if (items.putIfAbsent(descriptor.id, descriptor) != null) {
            throw ResourceException(
                category = ResourceErrorCategory.RESOLUTION_ERROR,
                message = "duplicate resource id: ${descriptor.id}",
            )
        }
    }

    override fun lookup(id: String): RegistryLookupResult {
        val descriptor = items[id] ?: return RegistryLookupResult.Missing
        val ttlMillis = descriptor.ttlMillis ?: return RegistryLookupResult.Hit(descriptor)
        return if (descriptor.createdAtMillis + ttlMillis <= nowMillis()) {
            RegistryLookupResult.Expired
        } else {
            RegistryLookupResult.Hit(descriptor)
        }
    }

    override fun remove(id: String) {
        items.remove(id)
    }

    override fun cleanup(nowMillis: Long) {
        items.entries.removeIf { (_, descriptor) ->
            descriptor.ttlMillis?.let { descriptor.createdAtMillis + it <= nowMillis } == true
        }
    }
}
